server/src/main/java/com/microsoft/java/bs/core/internal/services/BuildTargetService.java

```diff
@@ -469,6 +477,86 @@ public ScalacOptionsResult getBuildTargetScalacOptions(ScalacOptionsParams param
     return new ScalacOptionsResult(items);
   }
 
+  /**
+   * Run the test classes.
+   */
+  public TestResult buildTargetTest(TestParams params) {
+    TestResult testResult = new TestResult(StatusCode.OK);
+    testResult.setOriginId(params.getOriginId());
+    // running tests can trigger compilation that must be reported on
+    CompileProgressReporter compileProgressReporter = new CompileProgressReporter(client,
+            params.getOriginId(), getFullTaskPathMap());
+    Map<URI, Set<BuildTargetIdentifier>> groupedTargets =
+        groupBuildTargetsByRootDir(params.getTargets());
+    for (Map.Entry<URI, Set<BuildTargetIdentifier>> entry : groupedTargets.entrySet()) {
+      // TODO ideally BSP would have a jvmTestEnv style testkind for executing tests, not scala.
+      StatusCode statusCode;
+      if (TestParamsDataKind.SCALA_TEST.equals(params.getDataKind())) {
+        // ScalaTestParams is for a list of classes only
+        ScalaTestParams testParams = JsonUtils.toModel(params.getData(), ScalaTestParams.class);
+        Map<BuildTargetIdentifier, Map<String, Set<String>>> testClasses = new HashMap<>();
+        for (ScalaTestClassesItem testClassesItem : testParams.getTestClasses()) {
+          Map<String, Set<String>> classesMethods = new HashMap<>();
+          for (String classNames : testClassesItem.getClasses()) {
+            classesMethods.put(classNames, null);
+          }
+          testClasses.put(testClassesItem.getTarget(), classesMethods);
+        }
+        statusCode = connector.runTests(entry.getKey(), testClasses, testParams.getJvmOptions(),
+            params.getArguments(), null, client, params.getOriginId(),
+            compileProgressReporter);
+      } else if ("scala-test-suites-selection".equals(params.getDataKind())) {
+        // ScalaTestSuites is for a list of classes + methods
```

Comment:
Looks like `ScalaTestParams` doesn't supply the specific BuildTarget as well. Why do we only check it here?

Reply:
You mean `ScalaTestParams` _does_ supply the build target (inside each `ScalaTestClassesItem`) and `ScalaTestSuites` doesn't?
I'm not sure why `ScalaTestParams` does supply the build target and `ScalaTestSuites` doesn't - seems like an oversight.
Do you mean we should be filtering the `ScalaTestClassesItem` against `entry.getValue`?  You could be right about that.  Want me to add it?
Also - I've been force pushing every update as a single commit - do you prefer additional commits - does that make it easier to review changes?

Reply:
My bad! I didn't realize that `ScalaTestClassesItem` contains the build target. Sorry for the confusion.

> Do you mean we should be filtering the ScalaTestClassesItem against entry.getValue?

No, I'm not sure about that. I'm not clear about the spec here. Since both TestParam and ScalaTestClassesItem specifies the build target, which one wins when they are different?

> do you prefer additional commits 

Sure! Additional commits makes it easier to review. I can squash all the commits when merge.